### PR TITLE
feat(proposal-a): Phase 1 recall governance (Issue #569)

### LINKED ISSUE #569: [RFC] Proposal A Phase 1-4 Stack Chain 順序確認
**情況**：設計問題——涉及多個文件/子系統的更改

**為什麼**：在編寫程式碼之前，需要先確定方法。具體實施方案取決於答案。

---

### 現況說明

我正在處理 Issue #445 的 Proposal A（Dynamic Importance Feedback Signals）。Phase 1 ~ Phase 4 的實作已完成，但之前因為分支架構問題被關閉（PR #507/#505）。

目前我在 fork (`jlin53882/memory-lancedb-pro`) 上有以下分支：

| Branch | Phase | 最新 commit |
|--------|-------|------------|
| `feat/proposal-a-v3-clean` | Phase 1 | `649e90631b` |
| `feat/proposal-a-v3-configurable-v2` | Phase 3 | `4ce42d63be` |
| `feat/proposal-a-v3-tests` | Phase 4 | `52d822d080` |

三個 PR 都 targeting master，各自獨立會改同一批檔案衝突。

---

### Phase 1-4 實作範圍

| Phase | 內容 | 實作位置 |
|-------|------|---------|
| Phase 1 | agent_end hook + isRecallUsed() + importance delta 計算 | index.ts, reflection-slices.ts |
| Phase 2 | min_recall_count + bad_recall_count increment | 未實作 |
| Phase 3 | FeedbackConfigManager 參數化 | feedback-config.ts |
| Phase 4 | 單元測試覆蓋 | test/*.test.mjs |

---

#### 1. Stack Chain 的正確順序

根據你之前的建議，三個 PR 應以 stack chain 方式提交：

```
CortexReach/memory-lancedb-pro:master
  └── feat/proposal-a-v3-clean           (Phase 1) → PR
      └── feat/proposal-a-v3-configurable-v2  (Phase 3) → PR
          └── feat/proposal-a-v3-tests         (Phase 4) → PR #506
```

**請確認：**
- Phase 1 的 base 為 official master，對嗎？
- Phase 3 的 base 應為 Phase 1 branch？
- Phase 4 的 base 應為 Phase 3 branch，等 Phase 3 合併後再 rebase 到 master？

#### 2. Phase 1 → Phase 3 → Phase 4 的順序是對的嗎？

目前 Phase 2（min_recall_count + bad_recall_count increment）尚未實作。

**請確認：** 是否需要先實作 Phase 2？還是目前 Phase 1 → Phase 3 → Phase 4 的順序可以接受？

#### 3. 與 official master 同步

我的分支目前落後 official master。

**請確認：** 重建 stack chain 之前，是否需要先 rebase 到最新的 official master？

---

### 等待回饋

請確認以上三點。我會根據回覆重建正確的 stack chain，再陸續提交 PR。

---

Related to Issue #445

### PR DESCRIPTION
**Pull request by jlin53882**
## Summary

Phase 1 of Proposal A (recall governance): fixes critical bugs in the feedback loop for auto-recall memories, alignment of suppression and scoring thresholds, and natural usage detection.

This is a **clean rebuild** of PR #597/#698, cherry-picked onto current upstream master (f3be3a1).

See RFC: https://github.com/CortexReach/memory-lancedb-pro/issues/569

---

### All blockers from rwmjhb review are fixed:
- **F1**: Feedback defaults evaluate to 0 — fixed with proper nullish coalescing
- **MR1**: Feedback config added to plugin manifest schema
- **MR2**: before_prompt_build hook only fires when pendingRecall has entries

### Bug fixes applied:
1. **P0-1**: pendingRecall TTL cleanup (10 min max age)
2. **P0-3**: isRecallUsed() Summary path AND gate fixed
3. **P2**: suppression threshold aligned (>= 3 -> >= 2)
4. **P1**: Summary path independent of markers
5. **P1**: Scoring penalty threshold aligned (>= 2 -> >= 1)
6. **P1**: errorKeywords checked before usage heuristic
7. **P2**: TTL cleanup runs on both read and set paths

### New Files
- src/reflection-slices.ts: isRecallUsed function + usage tracking
- src/auto-capture-cleanup.ts: boilerplate wrapper stripping

### Config Changes
- openclaw.plugin.json: added eedback config schema

---

## Rebuild notes

This PR was **rebuilt from scratch** using cherry-pick onto current upstream master, ensuring all upstream changes are preserved:
- requestDimensions fix (#572)
- per-agent auto-recall control (#535)
- bulkStore batch write optimization
- Singleton State Management

Closes #569

### CODE REVIEW

**Comment by jlin53882** [comment]
## CI Status Note

部分 CI jobs 失敗，但這些失敗**與本 PR 修改無關**，為預先存在的環境問題：

- **`cli-smoke`**: 測試本身的 assertion bug（line 316），在 upstream/master 也會失敗
- **`core-regression` / `storage-and-schema` / `packaging-and-workflow`**: jiti mlly 依賴解析錯誤，本地 npm 環境問題，非本 PR 造成

本 PR 的 Phase 1 變更未觸發任何 CI 失敗。

**Comment by jlin53882** [comment]
## Unit Tests Added

新增 `test/is-recall-used.test.mjs` — 38 個單元測試，全部通過 ✅

覆蓋 `isRecallUsed()` 三條偵測路徑：

### 1. ID + Marker AND Logic（6 tests）
- ID 存在但無 marker → `false`
- ID 存在且有 marker → `true`
- Marker 存在但無 ID → `false`
- 多個 ID，其中一個匹配 → `true`
- 中文 marker（含「之前」）+ ID match

### 2. Summary Verbatim Path（5 tests）
- Summary verbatim 在 response 中 → `true`
- Summary 正好 9 字 → `false`（F3 >= 10 guard）
- Summary 正好 10 字 → `true`
- Whitespace/case-insensitive matching

### 3. Summary Reverse-Match（4 tests）
- Response 9-char snippet 在 summary → `false`（F3 guard）
- Response 10+ char snippet 在 summary → `true`
- First 50-char snippet 匹配邊界

### 4. Boundary Tests（7 tests）
- Response 正好 24 字 → `false`
- Response 正好 25 字 + ID+marker → `true`
- Summary 10-char boundary（uppercase/mixed-case）
- Response <= 24 字即使有 match → `false`

### 5. Edge Cases（12 tests）
- Empty arrays / whitespace-only IDs
- 多 summary 其中一個滿足條件
- 全部 ID 空字串 / mixed valid+empty
- CJK + 英文 cross-language 消極測試
- 獨立測試：injectedIds undefined + summary 有效 / injectedSummaries undefined + ID 有效

### 6. Mixed Paths（3 tests）
- ID fail + Summary succeed → `true`
- ID succeed + Summary fail → `true`
- Both fail → `false`

---

CI 失敗為預先存在的環境問題（非本 PR 造成）。

**Comment by AliceLJY** [comment]
@jlin53882 这个 PR body 已经写明是 #597 的 clean rebuild on master (f3be3a1)。建议把 #597 close 掉指向这里，让 review 集中在 #706 一条线上，不然恒哥两边盯起来很碎。

**Comment by jlin53882** [comment]
好的 我去把 #PR597 close

File: index.ts [hunk 1]
```diff
@@ -4243,15 +4515,12 @@ export function parsePluginConfig(value: unknown): PluginConfig {
...
-                ? ((cfg.recallPrefix as Record<string, unknown>).categoryField as string)
-                : undefined,
-          }
-        : undefined,
+    // Bug 3 fix: parse and return the feedback config block so deployments
+    // that specify custom feedback values actually take effect instead of
+    // falling back to hardcoded defaults.
+    feedback: typeof cfg.feedback === "object" && cfg.feedback !== null
+      ? { ...(cfg.feedback as Record<string, unknown>) }
+      : {},
```
[the remark this answers is not in the document]
Comment (jlin53882) [reply index.ts]: ✅ **Fixed —  regression restored**

commit cba8d5f — branch: 

Two changes applied:

1. **PluginConfig interface** (index.ts ~line 250):
   Added  to the interface definition with full JSDoc.

2. **parsePluginConfig mapping** (index.ts ~line 4610):
   Added  parsing to match the interface:
   ```ts
   recallPrefix: typeof cfg.recallPrefix === "object" \&\& cfg.recallPrefix !== null
     ? { ...(cfg.recallPrefix as Record<string, unknown>) }
     : {},
   ```

The code at ~line 2616 that reads  now works correctly. Deployments relying on custom metadata fields for `[category:scope]` labels are restored.

---

**Additional fixes found during concurrent Claude CLI adversarial review:**

3. **bulkStore + mdMirror isolation** — Auto-capture loop now uses bulkStore() once (1 lock instead of N). mdMirror moved after bulkStore succeeds; its errors are handled separately and do NOT trigger the store.store() fallback (prevents duplicate rows in LanceDB).

4. **fallback dedup** — The fallback loop (triggered when bulkStore fails) now re-dedups with vectorSearch before each individual store.store() call, ensuring dedup consistency across both paths.

**Comment by jlin53882** [comment]
## ✅ Codex Review 回覆 — 全部修復完成

**Commit:** `cba8d5f` on `proposal-a-phase1-v2`  
**Branch 已清理：** 無 graphify-out 雜質，僅含 PR #706 相關變更

---

### 回應 Codex Issue — `recallPrefix` Regression（P1）

**問題：** `parsePluginConfig` 移除了 `recallPrefix` mapping，但 `config.recallPrefix?.categoryField`（約 line 2616）仍在讀取，導致自訂 category 覆寫功能靜默失效。

**修復：**
1. **PluginConfig 介面**（~line 250）：補回 `recallPrefix?: { categoryField?: string }` + JSDoc 說明
2. **parsePluginConfig mapping**（~line 4610）：補回 parsing 邏輯

---

### Claude CLI 對抗性審查發現的額外問題

**Issue B/F — bulkStore fallback 缺 dedup 檢查**

風險：當 `bulkStore` 失敗進入 fallback 個別寫入時，若 `bulkStore` 部分 commit 了（entry 1 成功，entry 2 失敗），fallback 的 `store.store()` 會跳過 dedup 直接寫入，導致重複列。

**修復：** Fallback 迴圈在每次 `store.store()` 前重新執行 vectorSearch dedup 檢查。

**Issue E — mdMirror 錯誤會觸發 store.store() fallback**

風險：原實作中 `mdMirror` 在 `store.store()` 迴圈內，mdMirror 失敗時會直接執行到下一個 `store.store()`；但若整體 catch 捕獲，會觸發 fallback 迴圈，導致重複寫入。

**修復：** `mdMirror` 移到 `bulkStore` 成功後執行，各自錯誤獨立處理。mdMirror 失敗僅 log，不觸發 `store.store()` fallback。

---

### 修復確認

| 問題 | 來源 | 狀態 |
|------|------|------|
| recallPrefix regression | Codex | ✅ 已修復 |
| bulkStore fallback dedup | Claude | ✅ 已修復 |
| mdMirror isolation | Claude | ✅ 已修復 |
| boostOnUse ?? null ?? 0.05 | Codex | ✅ **非 bug**（設計行為） |
| Scoring loop 早停不還原 pendingRecall | Codex | ❓ 需完整 PR 範圍確認 |
